README.md
# Money
[![Build Status](https://travis-ci.org/liuggio/money.svg)](https://travis-ci.org/liuggio/money)

Elixir library for working with Money safer, easier, and fun,
is an interpretation of the Fowler's Money pattern in fun.prog.

> "If I had a dime for every time I've seen someone use FLOAT to store currency, I'd have $999.997634" -- [Bill Karwin](https://twitter.com/billkarwin/status/347561901460447232)

In short: You shouldn't represent monetary values by a float. Wherever
you need to represent money, use `Money`.

Documentation can be found at [https://hexdocs.pm/money/Money.html](https://hexdocs.pm/money/Money.html) on [HexDocs](https://hexdocs.pm)

## USAGE

```elixir
five_eur         = Money.new(500, :EUR)             # %Money{amount: 500, currency: :EUR}
ten_eur          = Money.add(five_eur, five_eur)    # %Money{amount: 10_00, currency: :EUR}
hundred_eur      = Money.multiply(ten_eur, 10)      # %Money{amount: 100_00, currency: :EUR}
ninety_nine_eur  = Money.subtract(hundred_eur, 1)   # %Money{amount: 99_00, currency: :EUR}
shares           = Money.divide(ninety_nine_eur, 2)
[%Money{amount: 50, currency: :EUR}, %Money{amount: 49, currency: :EUR}]

Money.equals?(five_eur, Money.new(500, :EUR)) # true
Money.zero?(five_eur);                        # false
Money.positive?(five_eur);                    # true

Money.Currency.symbol(:USD)                   # $
Money.Currency.symbol(Money.new(500, :AFN))   # ؋
Money.Currency.name(Money.new(500, :AFN))     # Afghani

Money.to_string(Money.new(500, :CNY))         # ¥ 5.00
Money.to_string(Money.new(1_234_56, :EUR), separator: ".", delimeter: ",", symbol: false)
"1.234,56"
Money.to_string(Money.new(1_234_56, :USD), fractional_unit: false)  # "$1,234"
```

### Money.Sigils

```elixir
# Sigils for Money
import Money.Sigils

iex> ~M[1000]USD
%Money{amount: 1000, currency: :USD}

# If you have a default currency configured (e.g. to GBP), you can do
iex> ~M[1000]
%Money{amount: 1000, currency: :GBP}
```

### Money.Currency

```elixir
# Currency convenience methods
import Money.Currency, only: [usd: 1, eur: 1, afn: 1]

iex> usd(100_00)
%Money{amount: 10000, currency: :USD}
iex> eur(100_00)
%Money{amount: 10000, currency: :EUR}
iex> afn(100_00)
%Money{amount: 10000, currency: :AFN}

Money.Currency.symbol(:USD)     # $
Money.Currency.symbol(afn(500)) # ؋
Money.Currency.name(afn(500))   # Afghani
Money.Currency.get(:AFN)        # %{name: "Afghani", symbol: "؋"}
```

### Money.Ecto.Type

Bring `Money` to your Ecto project.
The underlying database type is `integer`

```elixir
# migration
create table(:my_table) do
  add :amount, :integer
end

# model/schema
schema "my_table" do
  field :amount, Money.Ecto.Type
end
```

### Phoenix.HTML.Safe

Bring `Money` to your Phoenix project.
If you are using Phoenix, you can include money objects directly into your output and they will be correctly escaped.

```elixir
<b><%= Money.new(12345,67, :GBP) %></b>
```

## INSTALLATION

Money comes with no required dependencies.

Add the following to your `mix.exs`:

```elixir
def deps do
  [{:money, "~> 1.2.2"}]
end
```
then run [`mix deps.get`](http://elixir-lang.org/getting-started/mix-otp/introduction-to-mix).

## CONFIGURATION

You can set a default currency and global formatting preferences as follows:

```elixir
config :money,
  default_currency: :EUR,
  separator: ".",
  delimeter: ",",
  fractional_unit: true
  symbol: false,
  symbol_on_right: false,
  symbol_space: false
```

Then you don’t have to specify the currency. Also other config options will apply by default to all conversions.

```elixir
iex> amount = Money.new(1_234_50)
%Money{amount: 123450, currency: :EUR}
iex> Money.to_string(amount)
"1.234,50"
```

You can also pass formatting options to the `to_string` function.
Here is another example of formatting money:

```elixir
iex> amount = Money.new(1_234_50)
%Money{amount: 123450, currency: :EUR}
iex> Money.to_string(amount, symbol: true, symbol_on_right: true, symbol_space: true)
"1.234,50 €"
```

### Currency configuration
Sometimes you will need to setup different formatting options for each currency. You can do so by creating different
configuration for each currency you need to be applied on.
If some options are not found for the currency, it will fallback then to global and default options.

```elixir
config :money, :EUR
  separator: "_",
  symbol: true,
  symbol_on_right: true,
  symbol_space: true
```

Then:

```elixir
iex> amount = Money.new(1_234_50)
%Money{amount: 123450, currency: :EUR}
iex> Money.to_string(amount)
"1_234,50 €"
```

Of course, you can override this options on the `to_string` function when called:

```elixir
iex> amount = Money.new(1_234_50)
%Money{amount: 123450, currency: :EUR}
iex> Money.to_string(amount, symbol: true, symbol_on_right: false, symbol_space: false)
"€1_234,50"
```

## LICENSE

MIT License please see the [LICENSE](./LICENSE) file.


CHANGELOG.md
# Changelog

## 1.2.2
- Add Money.Config module for dealing with configuration tree (currency -> gloabal -> defaults)
- Doc for how-to add currency specific config

## 1.2.1

- Add Money.abs/1
- Add Money.neg/1
- Fix issues with Money.divide/2 and negative values [#57]
- Changes for clean compile on Elixir 1.4.1
- Changed sigil functions to macros for compile time expansion [#59]

## 1.2.0

- Add formatting option to hide fractional units [#47]
- Allow underscores in sigil [#48]

### 1.1.3

- Fix parsing amounts without a leading digit [#46]

### 1.1.2

- Bug fix: to_string with negative [#42]

### 1.1.1

- No changes

## 1.1.0

- Add formatting options [#36]
- Update parse to handle negative values [#39]
- Add support for Money.cast/1 for existing Money struct [#33]
- Add support for Money sigil ~M [#32]
- Add optional support for Phoenix.HTML.Safe [#29]

## 1.0.0

Initial release


test/money/config_test.exs
defmodule Money.ConfigTest do
  use ExUnit.Case

  alias Money.Config

  setup_all do
    # Global config
    Application.put_env(:money, :default_currency, :EUR)
    Application.put_env(:money, :separator, "_")
    Application.put_env(:money, :fractional_unit, false)
    Application.put_env(:money, :symbol, false)
    Application.put_env(:money, :symbol_on_right, false)
    Application.put_env(:money, :symbol_space, false)

    # EUR Currency config override
    Application.put_env(:money, :EUR, %{
      fractional_unit: true,
      symbol: true,
      symbol_on_right: true,
      symbol_space: true
    })

    on_exit fn ->
      Application.delete_env(:money, :default_currency)
      Application.delete_env(:money, :separator)
      Application.delete_env(:money, :delimeter)
      Application.delete_env(:money, :fractional_unit)
      Application.delete_env(:money, :symbol)
      Application.delete_env(:money, :symbol_on_right)
      Application.delete_env(:money, :symbol_space)
      Application.delete_env(:money, :EUR)
    end
  end

  describe "get/2" do
    test "returns EUR configured options" do
      assert Config.get(:fractional_unit, :EUR) == true
      assert Config.get(:symbol, :EUR) == true
      assert Config.get(:symbol_on_right, :EUR) == true
      assert Config.get(:symbol_space, :EUR) == true
    end

    test "fallbacks to global option if EUR has not the option configured" do
      assert Config.get(:separator, :EUR) == "_"
    end

    test "fallbacks to default option if EUR and global option is not configured" do
      assert Config.get(:delimeter, :EUR) == "."
    end
  end

  describe "get/1" do
    test "return the given global configured option" do
      assert Config.get(:default_currency) == :EUR
      assert Config.get(:separator) == "_"
      assert Config.get(:fractional_unit) == false
      assert Config.get(:symbol) == false
      assert Config.get(:symbol_on_right) == false
      assert Config.get(:symbol_space) == false
    end

    test "fallbacks to default option if option is not cofigured as global" do
      assert Config.get(:delimeter) == "."
    end
  end
end


mix.exs
defmodule Money.Mixfile do
  use Mix.Project

  @version "1.2.2"
  @github_url "https://github.com/liuggio/money"

  def project do
    [app: :money,
     name: "Money",
     version: @version,
     elixir: "~> 1.0",
     deps: deps(),
     source_url: "https://github.com/liuggio/money",
     docs: fn ->
       [source_ref: "v#{@version}",
        canonical: "http://hexdocs.pm/money",
        main: "Money",
        source_url: @github_url,
        extras: ["README.md", "CONTRIBUTING.md"]
       ]
     end,
     description: description(),
     package: package()]
  end

  def application do
    []
  end

  defp deps do
    [
      # Soft dependencies
      {:ecto, "~> 1.0 or ~> 2.0 or ~> 2.1", optional: true},
      {:phoenix_html, "~> 2.0", optional: true},

      # Code style
      {:credo, "~> 0.5", only: [:dev, :test]},

      # Docs
      {:ex_doc, "~> 0.14", only: [:dev, :docs]},
    ]
  end

  defp description do
    """
    Elixir library for working with Money safer, easier, and fun, is an interpretation of the Fowler's Money pattern in fun.prog.
    """
  end

  defp package do
    [
     maintainers: ["Giulio De Donato", "Andrew Timberlake"],
     contributors: ["Giulio De Donato", "Andrew Timberlake"],
     licenses: ["MIT"],
     links: %{"GitHub" => @github_url}]
  end
end


lib/money/config.ex
defmodule Money.Config do
  @moduledoc ~S"""
  Defines helper methods to retrieve Money configuration by currency.
  """

  @default_options %{
    symbol: true,
    symbol_on_right: false,
    symbol_space: false,
    fractional_unit: true,
    separator: ",",
    delimeter: "."
  }

  @spec get(atom, atom) :: binary | true | false
  @spec get(atom) :: binary | true | false

  @doc ~S"""
  Retrieve configuration options following the fallback tree for default options:
  - If `currency` argument is given it will try to retrive that `option` from the config about the given currency.
  - If this option is not found or no currency is given, then the global config `option` will be returned.
  - If no value for that `option` is found on the global config, then the default option is returned.
  """
  def get(option, currency),
    do: Application.get_env(:money, currency, [])[option] || get(option)
  def get(option),
    do: Application.get_env(:money, option, @default_options[option])
end


lib/money.ex
defmodule Money do
  import Kernel, except: [abs: 1]

  @moduledoc """
  Defines a `Money` struct along with convenience methods for working with currencies.

  ## Example:

      iex> money = Money.new(500, :USD)
      %Money{amount: 500, currency: :USD}
      iex> money = Money.add(money, 550)
      %Money{amount: 1050, currency: :USD}
      iex> Money.to_string(money)
      "$10.50"

  ## Configuration options

  You can set defaults in your Mix configuration to make working with `Money` a little easier.

  ## Configuration:

      config :money,
        default_currency: :EUR,  # this allows you to do Money.new(100)
        separator: ".",          # change the default thousands separator for Money.to_string
        delimiter: ",",          # change the default decimal delimeter for Money.to_string
        symbol: false            # don’t display the currency symbol in Money.to_string
        symbol_on_right: false,  # position the symbol
        symbol_space: false      # add a space between symbol and number
        fractional_unit: false   # don’t display the remainder or the delimeter

  ### Currency configuration
  Sometimes you will need to setup different formatting options for each currency. You can do so by creating different
  configuration for each currency you need to be applied on.
  If some options are not found for the currency, it will fallback then to global and default options.

      config :money, :EUR
        separator: "_",
        symbol: true,
        symbol_on_right: true,
        symbol_space: true
  """

  @type t :: %__MODULE__{
    amount: integer,
    currency: atom
  }

  defstruct amount: 0, currency: :USD

  alias Money.{Currency, Config}

  @spec new(integer) :: t
  @doc ~S"""
  Create a new `Money` struct using a default currency.
  The default currency can be set in the system Mix config.

  ## Example Config:

      config :money,
        default_currency: :USD

  ## Example:

      Money.new(123)
      %Money{amount: 123, currency: :USD}
  """
  def new(amount) do
    currency = Config.get(:default_currency)
    if currency do
      new(amount, currency)
    else
      raise ArgumentError, "to use Money.new/1 you must set a default currency in your application config."
    end
  end

  @spec new(integer, atom | String.t) :: t
  @doc """
  Create a new `Money` struct from currency sub-units (cents)

  ## Example:

      iex> Money.new(1_000_00, :USD)
      %Money{amount: 1_000_00, currency: :USD}
  """
  def new(int, currency) when is_integer(int),
    do: %Money{amount: int, currency: Currency.to_atom(currency)}

  @spec parse(String.t | float, atom | String.t, Keyword.t) :: {:ok, t}
  @doc ~S"""
  Parse a value into a `Money` type.

  The following options are available:

    - `separator` - default `","`, sets the separator for groups of thousands.
      "1,000"
    - `delimeter` - default `"."`, sets the decimal delimeter.
      "1.23"

  ## Examples:

      iex> Money.parse("$1,234.56", :USD)
      {:ok, %Money{amount: 123456, currency: :USD}}
      iex> Money.parse("1.234,56", :EUR, separator: ".", delimeter: ",")
      {:ok, %Money{amount: 123456, currency: :EUR}}
      iex> Money.parse("1.234,56", :WRONG)
      :error
      iex> Money.parse(1_234.56, :USD)
      {:ok, %Money{amount: 123456, currency: :USD}}
      iex> Money.parse(-1_234.56, :USD)
      {:ok, %Money{amount: -123456, currency: :USD}}
  """
  def parse(value, currency \\ nil, opts \\ [])
  def parse(value, nil, opts) do
    currency = Config.get(:default_currency)
    if currency do
      parse(value, currency, opts)
    else
      raise ArgumentError, "to use Money.new/1 you must set a default currency in your application config."
    end
  end
  def parse(str, currency, opts) when is_binary(str) do
    try do
      {_separator, delimeter} = get_parse_options(currency, opts)
      value = str
      |> prepare_parse_string(delimeter)
      |> add_missing_leading_digit
      case Float.parse(value) do
        {float, _} -> parse(float, currency, [])
        :error -> :error
      end
    rescue
      _ -> :error
    end
  end
  def parse(float, currency, _opts) when is_float(float) do
    {:ok, new(round(float * Currency.sub_units_count!(currency)), currency)}
  end

  defp prepare_parse_string(characters, delimeter, acc \\ [])
  defp prepare_parse_string([], _delimeter, acc),
    do: Enum.reverse(acc) |> Enum.join
  defp prepare_parse_string(["-" | tail], delimeter, acc),
    do: prepare_parse_string(tail, delimeter, ["-" | acc])
  defp prepare_parse_string(["0" | tail], delimeter, acc),
    do: prepare_parse_string(tail, delimeter, ["0" | acc])
  defp prepare_parse_string(["1" | tail], delimeter, acc),
    do: prepare_parse_string(tail, delimeter, ["1" | acc])
  defp prepare_parse_string(["2" | tail], delimeter, acc),
    do: prepare_parse_string(tail, delimeter, ["2" | acc])
  defp prepare_parse_string(["3" | tail], delimeter, acc),
    do: prepare_parse_string(tail, delimeter, ["3" | acc])
  defp prepare_parse_string(["4" | tail], delimeter, acc),
    do: prepare_parse_string(tail, delimeter, ["4" | acc])
  defp prepare_parse_string(["5" | tail], delimeter, acc),
    do: prepare_parse_string(tail, delimeter, ["5" | acc])
  defp prepare_parse_string(["6" | tail], delimeter, acc),
    do: prepare_parse_string(tail, delimeter, ["6" | acc])
  defp prepare_parse_string(["7" | tail], delimeter, acc),
    do: prepare_parse_string(tail, delimeter, ["7" | acc])
  defp prepare_parse_string(["8" | tail], delimeter, acc),
    do: prepare_parse_string(tail, delimeter, ["8" | acc])
  defp prepare_parse_string(["9" | tail], delimeter, acc),
    do: prepare_parse_string(tail, delimeter, ["9" | acc])
  defp prepare_parse_string([delimeter | tail], delimeter, acc),
    do: prepare_parse_string(tail, delimeter, ["." | acc])
  defp prepare_parse_string([_head | tail], delimeter, acc),
    do: prepare_parse_string(tail, delimeter, acc)

  defp prepare_parse_string(string, delimeter, _acc),
    do: prepare_parse_string(String.codepoints(string), delimeter)

  defp add_missing_leading_digit(<< "-." >> <> tail),
    do: "-0." <> tail
  defp add_missing_leading_digit(<< "." >> <> tail),
    do: "0." <> tail
  defp add_missing_leading_digit(str), do: str

  @spec parse(String.t | float, atom | String.t, Keyword.t) :: t
  @doc ~S"""
  Parse a value into a `Money` type.
  Similar to `parse/3` but returns a `%Money{}` or raises an error if parsing fails.

  ## Example:

      iex> Money.parse!("1,234.56", :USD)
      %Money{amount: 123456, currency: :USD}
      iex> Money.parse!("wrong", :USD)
      ** (ArgumentError) unable to parse "wrong"
  """
  def parse!(value, currency \\ nil, opts \\ []) do
    case parse(value, currency, opts) do
      {:ok, money} -> money
      :error -> raise ArgumentError, "unable to parse #{inspect(value)}"
    end
  end

  @spec compare(t, t) :: t
  @doc ~S"""
  Compares two `Money` structs with each other.
  They must each be of the same currency and then their amounts are compared

  ## Example:

      iex> Money.compare(Money.new(100, :USD), Money.new(100, :USD))
      0
      iex> Money.compare(Money.new(100, :USD), Money.new(101, :USD))
      -1
      iex> Money.compare(Money.new(101, :USD), Money.new(100, :USD))
      1
  """
  def compare(%Money{currency: cur} = a, %Money{currency: cur} = b) do
    case a.amount - b.amount do
                    x when x >  0 -> 1
                    x when x <  0 -> -1
                    x when x == 0 -> 0
    end
  end
  def compare(a, b), do: fail_currencies_must_be_equal(a, b)

  @spec zero?(t) :: boolean
  @doc ~S"""
  Returns true if the amount of a `Money` struct is zero

  ## Example:

      iex> Money.zero?(Money.new(0, :USD))
      true
      iex> Money.zero?(Money.new(1, :USD))
      false
  """
  def zero?(%Money{amount: amount}) do
    amount == 0
  end

  @spec positive?(t) :: boolean
  @doc ~S"""
  Returns true if the amount of a `Money` is greater than zero

  ## Example:

      iex> Money.positive?(Money.new(0, :USD))
      false
      iex> Money.positive?(Money.new(1, :USD))
      true
      iex> Money.positive?(Money.new(-1, :USD))
      false
  """
  def positive?(%Money{amount: amount}) do
    amount > 0
  end

  @spec negative?(t) :: boolean
  @doc ~S"""
  Returns true if the amount of a `Money` is less than zero

  ## Example:

      iex> Money.negative?(Money.new(0, :USD))
      false
      iex> Money.negative?(Money.new(1, :USD))
      false
      iex> Money.negative?(Money.new(-1, :USD))
      true
  """
  def negative?(%Money{amount: amount}) do
    amount < 0
  end

  @spec equals?(t, t) :: boolean
  @doc ~S"""
  Returns true if two `Money` of the same currency have the same amount

  ## Example:

      iex> Money.equals?(Money.new(100, :USD), Money.new(100, :USD))
      true
      iex> Money.equals?(Money.new(101, :USD), Money.new(100, :USD))
      false
  """
  def equals?(%Money{amount: amount, currency: cur}, %Money{amount: amount, currency: cur}), do: true
  def equals?(%Money{currency: cur}, %Money{currency: cur}), do: false
  def equals?(a, b), do: fail_currencies_must_be_equal(a, b)

  @spec neg(t) :: t
  @doc ~S"""
  Returns a `Money` with the amount negated.

  ## Examples:

      iex> Money.new(100, :USD) |> Money.neg
      %Money{amount: -100, currency: :USD}
      iex> Money.new(-100, :USD) |> Money.neg
      %Money{amount: 100, currency: :USD}
  """
  def neg(%Money{amount: amount, currency: cur}),
    do: %Money{amount: -amount, currency: cur}

  @spec abs(t) :: t
  @doc ~S"""
  Returns a `Money` with the arithmetical absolute of the amount.

  ## Examples:

      iex> Money.new(-100, :USD) |> Money.abs
      %Money{amount: 100, currency: :USD}
      iex> Money.new(100, :USD) |> Money.abs
      %Money{amount: 100, currency: :USD}
  """
  def abs(%Money{amount: amount, currency: cur}),
    do: %Money{amount: Kernel.abs(amount), currency: cur}

  @spec add(t, t | integer | float) :: t
  @doc ~S"""
  Adds two `Money` together or an integer (cents) amount to a `Money`

  ## Example:

      iex> Money.add(Money.new(100, :USD), Money.new(50, :USD))
      %Money{amount: 150, currency: :USD}
      iex> Money.add(Money.new(100, :USD), 50)
      %Money{amount: 150, currency: :USD}
      iex> Money.add(Money.new(100, :USD), 5.55)
      %Money{amount: 655, currency: :USD}
  """
  def add(%Money{amount: a, currency: cur}, %Money{amount: b, currency: cur}),
    do: Money.new(a + b, cur)
  def add(%Money{amount: amount, currency: cur}, addend) when is_integer(addend),
    do: Money.new(amount + addend, cur)
  def add(%Money{} = m, addend) when is_float(addend),
    do: add(m, round(addend * 100))
  def add(a, b), do: fail_currencies_must_be_equal(a, b)

  @spec subtract(t, t | integer | float) :: t
  @doc ~S"""
  Subtracts one `Money` from another or an integer (cents) from a `Money`

  ## Example:

      iex> Money.subtract(Money.new(150, :USD), Money.new(50, :USD))
      %Money{amount: 100, currency: :USD}
      iex> Money.subtract(Money.new(150, :USD), 50)
      %Money{amount: 100, currency: :USD}
      iex> Money.subtract(Money.new(150, :USD), 1.25)
      %Money{amount: 25, currency: :USD}
  """
  def subtract(%Money{amount: a, currency: cur}, %Money{amount: b, currency: cur}),
    do: Money.new(a - b, cur)
  def subtract(%Money{amount: a, currency: cur}, subtractend) when is_integer(subtractend),
    do: Money.new(a - subtractend, cur)
  def subtract(%Money{} = m, subtractend) when is_float(subtractend),
    do: subtract(m, round(subtractend * 100))
  def subtract(a, b), do: fail_currencies_must_be_equal(a, b)

  @spec multiply(t, integer | float) :: t
  @doc ~S"""
  Multiplies a `Money` by an amount

  ## Example:
      iex> Money.multiply(Money.new(100, :USD), 10)
      %Money{amount: 1000, currency: :USD}
      iex> Money.multiply(Money.new(100, :USD), 1.5)
      %Money{amount: 150, currency: :USD}
  """
  def multiply(%Money{amount: amount, currency: cur}, multiplier) when is_integer(multiplier),
    do: Money.new(amount * multiplier, cur)
  def multiply(%Money{amount: amount, currency: cur}, multiplier) when is_float(multiplier),
    do: Money.new(round(amount * multiplier), cur)

  @spec divide(t, integer) :: [t]
  @doc ~S"""
  Divides up `Money` by an amount

  ## Example:
      iex> Money.divide(Money.new(100, :USD), 2)
      [%Money{amount: 50, currency: :USD}, %Money{amount: 50, currency: :USD}]
      iex> Money.divide(Money.new(101, :USD), 2)
      [%Money{amount: 51, currency: :USD}, %Money{amount: 50, currency: :USD}]
  """
  def divide(%Money{amount: amount, currency: cur}, denominator) when is_integer(denominator) do
    value = div(amount, denominator)
    rem   = rem(amount, denominator)
    do_divide(cur, value, rem, denominator, [])
  end

  defp do_divide(_currency, _value, _rem, 0, acc), do: acc |> Enum.reverse
  defp do_divide(currency, value, 0, count, acc) do
    count = decrement_abs(count)
    acc   = [new(value, currency) | acc]
    do_divide(currency, value, 0, count, acc)
  end
  defp do_divide(currency, value, rem, count, acc) do
    rem   = decrement_abs(rem)
    count = decrement_abs(count)
    acc   = [new(increment_abs(value), currency) | acc]
    do_divide(currency, value, rem, count, acc)
  end

  defp increment_abs(n) when n >= 0, do: n + 1
  defp increment_abs(n) when n < 0, do: n - 1
  defp decrement_abs(n) when n >= 0, do: n - 1
  defp decrement_abs(n) when n < 0, do: n + 1

  @spec to_string(t, Keyword.t) :: String.t
  @doc ~S"""
  Converts a `Money` struct to a string representation

  The following options are available:

    - `separator` - default `","`, sets the separator for groups of thousands.
      "1,000"
    - `delimeter` - default `"."`, sets the decimal delimeter.
      "1.23"
    - `symbol` - default `true`, sets whether to display the currency symbol or not.
    - `symbol_on_right` - default `false`, display the currency symbol on the right of the number, eg: 123.45€
    - `symbol_space` - default `false`, add a space between currency symbol and number, eg: € 123,45 or 123.45 €
    - `fractional_unit` - default `true`, show the remaining units after the delimeter

  ## Example:

      iex> Money.to_string(Money.new(123456, :GBP))
      "£1,234.56"
      iex> Money.to_string(Money.new(123456, :EUR), separator: ".", delimeter: ",")
      "€1.234,56"
      iex> Money.to_string(Money.new(123456, :EUR), symbol: false)
      "1,234.56"
      iex> Money.to_string(Money.new(123456, :EUR), symbol: false, separator: "")
      "1234.56"
      iex> Money.to_string(Money.new(123456, :EUR), fractional_unit: false)
      "€1,234"

  It can also be interpolated (It implements the String.Chars protocol)
  To control the formatting, you can use the above options in your config,
  more information is in the introduction to `Money`

  ## Example:

      iex> "Total: #{Money.new(100_00, :USD)}"
      "Total: $100.00"
  """
  def to_string(%Money{}=money, opts \\ []) do
    {separator, delimeter, symbol, symbol_on_right, symbol_space, fractional_unit} = get_display_options(money, opts)
    number = format_number(money, separator, delimeter, fractional_unit, money)
    sign = if negative?(money), do: "-"
    space = if symbol_space, do: " "

    parts = if symbol_on_right do
              [sign, number, space, symbol]
            else
              [symbol, space, sign, number]
            end
    parts |> Enum.join |> String.trim_leading
  end

  defp format_number(%Money{amount: amount}, separator, delimeter, fractional_unit, money) do
    exponent = Currency.exponent(money)
    sub_units_count = Currency.sub_units_count!(money)
    [super_unit | sub_unit] = Kernel.abs(amount/sub_units_count) |> :erlang.float_to_binary(decimals: exponent) |> String.split(".")
    super_unit = super_unit |> reverse_group(3) |> Enum.join(separator)
    if fractional_unit && sub_unit != [] do
      [super_unit, sub_unit] |> Enum.join(delimeter)
    else
      super_unit
    end
  end

  defp get_display_options(m, opts) do
    {separator, delimeter} = get_parse_options(m.currency, opts)

    default_symbol = Config.get(:symbol, m.currency)
    default_symbol_on_right = Config.get(:symbol_on_right, m.currency)
    default_symbol_space = Config.get(:symbol_space, m.currency)
    default_fractional_unit = Config.get(:fractional_unit, m.currency)

    symbol = if Keyword.get(opts, :symbol, default_symbol), do: Currency.symbol(m), else: ""
    symbol_on_right = Keyword.get(opts, :symbol_on_right, default_symbol_on_right)
    symbol_space = Keyword.get(opts, :symbol_space, default_symbol_space)
    fractional_unit = Keyword.get(opts, :fractional_unit, default_fractional_unit)

    {separator, delimeter, symbol, symbol_on_right, symbol_space, fractional_unit}
  end

  defp get_parse_options(currency, opts) do
    default_separator = Config.get(:separator, currency)
    default_delimeter = Config.get(:delimeter, currency)

    separator = Keyword.get(opts, :separator, default_separator)
    delimeter = Keyword.get(opts, :delimeter, default_delimeter)

    {separator, delimeter}
  end

  defp fail_currencies_must_be_equal(a, b) do
    raise ArgumentError, message: "Currency of #{a.currency} must be the same as #{b.currency}"
  end

  defp reverse_group(str, count) when is_binary(str) do
    reverse_group(str, Kernel.abs(count), [])
  end
  defp reverse_group("", _count, list) do
    list
  end
  defp reverse_group(str, count, list) do
    {first, last} = String.split_at(str, -count)
    reverse_group(first, count, [last | list])
  end

  defimpl String.Chars do
    def to_string(%Money{} = m) do
      Money.to_string(m)
    end
  end
end
